privacyidea/lib/monitoringmodules/base.py

```diff
@@ -0,0 +1,98 @@
+# License as published by the Free Software Foundation; either
+# version 3 of the License, or any later version.
+#
+# This code is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+#
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+__doc__ = """This is the BaseClass for monitoring statistics data.
+
+The monitoring database is a time series database. It is supposed to store
+tuples of (timestamp, status key, status value)
+
+A monitoring module needs to provide the possibility to write new data,
+return all available keys, return the last data. 
+"""
+import logging
+log = logging.getLogger(__name__)
+from privacyidea.lib.log import log_with
+
+
+class Monitoring(object):
+
+    def __init__(self, config=None):
+        pass
+
+    def get_keys(self):
+        """
+        Return a list of the available statistic keys.
+
+        :return: list of identifiers
+        """
+        return []
+
+    def get_values(self, stats_key, start_timestamp=None, end_timestamp=None, date_strings=False):
+        """
+        Return a list of tuples of (timestamp, value) for the requested stats_key.
+
+        :param stats_key: Identifier of the stats
+        :param start_timestamp: utc start of the time frame
+        :type start_timestamp: datetime
+        :param end_timestamp:  utc end of the time frame
+        :type end_timestamp: datetime
+        :param date_strings: If set to True, the timestamps are returned as strings. Otherwise as python timestamp
+        :return:
+        """
+        return []
+
+    def get_last_value(self, stats_key):
+        """
+        returns the last value of the given stats_key in time.
+        :param stats_key: The identifier of the stats
+        :return: a string value.
+        """
+        pass
+
+    def delete(self, stats_key, start_timestamp, end_timestamp):
+        """
+        Delete all entries of the stats_key for the given time frame.
+
+        :param stats_key: Identifier of the stats
+        :param start_timestamp: utc beginning of the time frame
+        :type start_timestamp: datetime
+        :param end_timestamp: utc end of the time frame
+        :type end_timestamp: datetime
+        :return: number of deleted entries
+        """
+        return 0
+
+    def add_value(self, stats_key, stats_value, utc_timestamp, reset_values=False):
+        """
+        This method adds a measurement point to the statistics key "stats_key".
+        If reset_values is set to True, all old values of this stats_key are deleted.
+
+        :param stats_key: Identifier of the stats
+        :param stats_value: measured value
+        :param utc_timestamp: the timestamp of the measurement
```

Comment:
we should clarify whether this is a timezone-aware datetime or not. See my comment on the SQLStats implementation of ``add_value`` later :)

Reply:
done in fa76b0b

Reply:
:+1: